tokio/src/runtime/time/mod.rs

```diff
@@ -319,23 +319,59 @@ impl Handle {
             now = lock.elapsed();
         }
 
-        while let Some(entry) = lock.poll(now) {
-            debug_assert!(unsafe { entry.is_pending() });
-
-            // SAFETY: We hold the driver lock, and just removed the entry from any linked lists.
-            if let Some(waker) = unsafe { entry.fire(Ok(())) } {
-                waker_list.push(waker);
-
-                if !waker_list.can_push() {
-                    // Wake a batch of wakers. To avoid deadlock, we must do this with the lock temporarily dropped.
-                    drop(lock);
-
-                    waker_list.wake_all();
-
-                    lock = self.inner.lock_sharded_wheel(id);
+        while let Some(expiration) = lock.poll(now) {
+            lock.set_elapsed(expiration.deadline);
+            // Note that we need to take _all_ of the entries off the list before
+            // processing any of them. This is important because it's possible that
+            // those entries might need to be reinserted into the same slot.
+            //
+            // This happens only on the highest level, when an entry is inserted
+            // more than MAX_DURATION into the future. When this happens, we wrap
+            // around, and process some entries a multiple of MAX_DURATION before
+            // they actually need to be dropped down a level. We then reinsert them
+            // back into the same position; we must make sure we don't then process
+            // those entries again or we'll end up in an infinite loop.
+            let unguarded_list = lock.take_entries(&expiration);
+            // It is critical for `GuardedLinkedList` safety that the guard node is
+            // pinned in memory and is not dropped until the guarded list is dropped.
+            let guard = TimerShared::new(id);
+            pin!(guard);
+            let guard_ptr = NonNull::from(guard.as_ref().get_ref());
+            // GuardedLinkedList ensures that the concurrent drop of Entry in this slot is safe.
+            let mut guarded_list = unguarded_list.into_guarded(guard_ptr);
```

Comment:
Can we move the call to `lock.take_entries` so that we immediately call `into_guarded`?

Reply:
I add the method `get_waiters_list` to replace it. Is this design reasonable enough for us?